From 5fc6c19370c60065188c9ce5d77987d231c1d434 Mon Sep 17 00:00:00 2001
From: Daniel Watkins <daniel.watkins@canonical.com>
Date: Wed, 18 Feb 2015 18:10:15 +0000
Subject: [PATCH 4/8] Set an explicit timeout when fetching CloudStack
 passwords.

---
 cloudinit/sources/DataSourceCloudStack.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/cloudinit/sources/DataSourceCloudStack.py b/cloudinit/sources/DataSourceCloudStack.py
index 211893573235..bf20798c30dc 100644
--- a/cloudinit/sources/DataSourceCloudStack.py
+++ b/cloudinit/sources/DataSourceCloudStack.py
@@ -143,6 +143,7 @@ class DataSourceCloudStack(sources.DataSource):
             # so we use the socket directly to read off the password.
             conn = http_client.HTTPConnection(self.vr_addr, 8080)
             conn.request('GET', '', headers={'DomU_Request': req_string})
+            conn.sock.settimeout(30)
             output = conn.sock.recv(1024).decode('utf-8').strip()
             conn.close()
             return output
-- 
2.11.0

