diff -uNr cloud-init-0.7.5__orig/cloudinit/distros/rhel.py cloud-init-0.7.5/cloudinit/distros/rhel.py
--- cloud-init-0.7.5__orig/cloudinit/distros/rhel.py	2014-04-01 19:26:07.000000000 +0100
+++ cloud-init-0.7.5/cloudinit/distros/rhel.py	2014-08-13 01:45:26.485837406 +0100
@@ -103,6 +103,7 @@
         (dist, vers) = util.system_info()['dist'][:2]
         major = (int)(vers.split('.')[0])
         return ((dist.startswith('Red Hat Enterprise Linux') and major >= 7)
+                or (dist.startswith('CentOS Linux') and major >= 7)
                 or (dist.startswith('Fedora') and major >= 18))
 
     def apply_locale(self, locale, out_fn=None):
